// playcanvas/build/playcanvas.dbg/src/extras/gizmo/constants.js
/**
 * Local coordinate space.
 *
 * @type {string}
 * @category Gizmo
 */ var GIZMOSPACE_LOCAL = 'local';
/**
 * World coordinate space.
 *
 * @type {string}
 * @category Gizmo
 */ var GIZMOSPACE_WORLD = 'world';
/**
 * Gizmo axis for the line X.
 *
 * @type {string}
 */ var GIZMOAXIS_X = 'x';
/**
 * Gizmo axis for the line Y.
 *
 * @type {string}
 */ var GIZMOAXIS_Y = 'y';
/**
 * Gizmo axis for the line Z.
 *
 * @type {string}
 */ var GIZMOAXIS_Z = 'z';
/**
 * Gizmo axis for the plane YZ.
 *
 * @type {string}
 */ var GIZMOAXIS_YZ = 'yz';
/**
 * Gizmo axis for the plane XZ.
 *
 * @type {string}
 */ var GIZMOAXIS_XZ = 'xz';
/**
 * Gizmo axis for the plane XY.
 *
 * @type {string}
 */ var GIZMOAXIS_XY = 'xy';
/**
 * Gizmo axis for all directions XYZ.
 *
 * @type {string}
 */ var GIZMOAXIS_XYZ = 'xyz';
/**
 * Gizmo axis for facing the camera (facing the camera).
 *
 * @type {string}
 */ var GIZMOAXIS_FACE = 'face';

export { GIZMOAXIS_FACE, GIZMOAXIS_X, GIZMOAXIS_XY, GIZMOAXIS_XYZ, GIZMOAXIS_XZ, GIZMOAXIS_Y, GIZMOAXIS_YZ, GIZMOAXIS_Z, GIZMOSPACE_LOCAL, GIZMOSPACE_WORLD };
