// playcanvas/build/playcanvas.prf/src/core/constants.js
var TRACEID_RENDER_FRAME = 'RenderFrame';
var TRACEID_RENDER_FRAME_TIME = 'RenderFrameTime';
var TRACEID_RENDER_PASS = 'RenderPass';
var TRACEID_RENDER_PASS_DETAIL = 'RenderPassDetail';
var TRACEID_RENDER_ACTION = 'RenderAction';
var TRACEID_RENDER_TARGET_ALLOC = 'RenderTargetAlloc';
var TRACEID_TEXTURE_ALLOC = 'TextureAlloc';
var TRACEID_SHADER_ALLOC = 'ShaderAlloc';
var TRACEID_SHADER_COMPILE = 'ShaderCompile';
var TRACEID_VRAM_TEXTURE = 'VRAM.Texture';
var TRACEID_VRAM_VB = 'VRAM.Vb';
var TRACEID_VRAM_IB = 'VRAM.Ib';
var TRACEID_VRAM_SB = 'VRAM.Sb';
var TRACEID_BINDGROUP_ALLOC = 'BindGroupAlloc';
var TRACEID_BINDGROUPFORMAT_ALLOC = 'BindGroupFormatAlloc';
var TRACEID_RENDERPIPELINE_ALLOC = 'RenderPipelineAlloc';
var TRACEID_COMPUTEPIPELINE_ALLOC = 'ComputePipelineAlloc';
var TRACEID_PIPELINELAYOUT_ALLOC = 'PipelineLayoutAlloc';
var TRACE_ID_ELEMENT = 'Element';
var TRACEID_TEXTURES = 'Textures';
var TRACEID_RENDER_QUEUE = 'RenderQueue';
var TRACEID_GPU_TIMINGS = 'GpuTimings';

export { TRACEID_BINDGROUPFORMAT_ALLOC, TRACEID_BINDGROUP_ALLOC, TRACEID_COMPUTEPIPELINE_ALLOC, TRACEID_GPU_TIMINGS, TRACEID_PIPELINELAYOUT_ALLOC, TRACEID_RENDERPIPELINE_ALLOC, TRACEID_RENDER_ACTION, TRACEID_RENDER_FRAME, TRACEID_RENDER_FRAME_TIME, TRACEID_RENDER_PASS, TRACEID_RENDER_PASS_DETAIL, TRACEID_RENDER_QUEUE, TRACEID_RENDER_TARGET_ALLOC, TRACEID_SHADER_ALLOC, TRACEID_SHADER_COMPILE, TRACEID_TEXTURES, TRACEID_TEXTURE_ALLOC, TRACEID_VRAM_IB, TRACEID_VRAM_SB, TRACEID_VRAM_TEXTURE, TRACEID_VRAM_VB, TRACE_ID_ELEMENT };
