// playcanvas/build/playcanvas.prf/src/framework/anim/controller/constants.js
var ANIM_INTERRUPTION_NONE = 'NONE';
var ANIM_INTERRUPTION_PREV = 'PREV_STATE';
var ANIM_INTERRUPTION_NEXT = 'NEXT_STATE';
var ANIM_INTERRUPTION_PREV_NEXT = 'PREV_STATE_NEXT_STATE';
var ANIM_INTERRUPTION_NEXT_PREV = 'NEXT_STATE_PREV_STATE';
var ANIM_GREATER_THAN = 'GREATER_THAN';
var ANIM_LESS_THAN = 'LESS_THAN';
var ANIM_GREATER_THAN_EQUAL_TO = 'GREATER_THAN_EQUAL_TO';
var ANIM_LESS_THAN_EQUAL_TO = 'LESS_THAN_EQUAL_TO';
var ANIM_EQUAL_TO = 'EQUAL_TO';
var ANIM_NOT_EQUAL_TO = 'NOT_EQUAL_TO';
var ANIM_PARAMETER_INTEGER = 'INTEGER';
var ANIM_PARAMETER_FLOAT = 'FLOAT';
var ANIM_PARAMETER_BOOLEAN = 'BOOLEAN';
var ANIM_PARAMETER_TRIGGER = 'TRIGGER';
var ANIM_BLEND_1D = '1D';
var ANIM_BLEND_2D_DIRECTIONAL = '2D_DIRECTIONAL';
var ANIM_BLEND_2D_CARTESIAN = '2D_CARTESIAN';
var ANIM_BLEND_DIRECT = 'DIRECT';
var ANIM_STATE_START = 'START';
var ANIM_STATE_END = 'END';
var ANIM_STATE_ANY = 'ANY';
var ANIM_CONTROL_STATES = [
		ANIM_STATE_START,
		ANIM_STATE_END,
		ANIM_STATE_ANY
];
var ANIM_LAYER_OVERWRITE = 'OVERWRITE';
var ANIM_LAYER_ADDITIVE = 'ADDITIVE';

export { ANIM_BLEND_1D, ANIM_BLEND_2D_CARTESIAN, ANIM_BLEND_2D_DIRECTIONAL, ANIM_BLEND_DIRECT, ANIM_CONTROL_STATES, ANIM_EQUAL_TO, ANIM_GREATER_THAN, ANIM_GREATER_THAN_EQUAL_TO, ANIM_INTERRUPTION_NEXT, ANIM_INTERRUPTION_NEXT_PREV, ANIM_INTERRUPTION_NONE, ANIM_INTERRUPTION_PREV, ANIM_INTERRUPTION_PREV_NEXT, ANIM_LAYER_ADDITIVE, ANIM_LAYER_OVERWRITE, ANIM_LESS_THAN, ANIM_LESS_THAN_EQUAL_TO, ANIM_NOT_EQUAL_TO, ANIM_PARAMETER_BOOLEAN, ANIM_PARAMETER_FLOAT, ANIM_PARAMETER_INTEGER, ANIM_PARAMETER_TRIGGER, ANIM_STATE_ANY, ANIM_STATE_END, ANIM_STATE_START };
