// playcanvas/build/playcanvas.prf/src/framework/xr/constants.js
var XRTYPE_INLINE = 'inline';
var XRTYPE_VR = 'immersive-vr';
var XRTYPE_AR = 'immersive-ar';
var XRSPACE_VIEWER = 'viewer';
var XRSPACE_LOCAL = 'local';
var XRSPACE_LOCALFLOOR = 'local-floor';
var XRSPACE_BOUNDEDFLOOR = 'bounded-floor';
var XRSPACE_UNBOUNDED = 'unbounded';
var XRTARGETRAY_GAZE = 'gaze';
var XRTARGETRAY_SCREEN = 'screen';
var XRTARGETRAY_POINTER = 'tracked-pointer';
var XREYE_NONE = 'none';
var XREYE_LEFT = 'left';
var XREYE_RIGHT = 'right';
var XRHAND_NONE = 'none';
var XRHAND_LEFT = 'left';
var XRHAND_RIGHT = 'right';
var XRTRACKABLE_POINT = 'point';
var XRTRACKABLE_PLANE = 'plane';
var XRTRACKABLE_MESH = 'mesh';
var XRDEPTHSENSINGUSAGE_CPU = 'cpu-optimized';
var XRDEPTHSENSINGUSAGE_GPU = 'gpu-optimized';
var XRDEPTHSENSINGFORMAT_L8A8 = 'luminance-alpha';
var XRDEPTHSENSINGFORMAT_R16U = 'unsigned-short';
var XRDEPTHSENSINGFORMAT_F32 = 'float32';

export { XRDEPTHSENSINGFORMAT_F32, XRDEPTHSENSINGFORMAT_L8A8, XRDEPTHSENSINGFORMAT_R16U, XRDEPTHSENSINGUSAGE_CPU, XRDEPTHSENSINGUSAGE_GPU, XREYE_LEFT, XREYE_NONE, XREYE_RIGHT, XRHAND_LEFT, XRHAND_NONE, XRHAND_RIGHT, XRSPACE_BOUNDEDFLOOR, XRSPACE_LOCAL, XRSPACE_LOCALFLOOR, XRSPACE_UNBOUNDED, XRSPACE_VIEWER, XRTARGETRAY_GAZE, XRTARGETRAY_POINTER, XRTARGETRAY_SCREEN, XRTRACKABLE_MESH, XRTRACKABLE_PLANE, XRTRACKABLE_POINT, XRTYPE_AR, XRTYPE_INLINE, XRTYPE_VR };
